extensions/amp-base-carousel/1.0/scroller.js

```diff
@@ -0,0 +1,231 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import * as Preact from '../../../src/preact';
+import * as styles from './base-carousel.css';
+import {WithAmpContext} from '../../../src/preact/context';
+import {debounce} from '../../../src/utils/rate-limit';
+import {mod} from '../../../src/utils/math';
+import {useLayoutEffect, useRef} from '../../../src/preact';
+
+/**
+ * How long to wait prior to resetting the scrolling position after the last
+ * scroll event. Ideally this should be low, so that once the user stops
+ * scrolling, things are immediately centered again. Since there can be some
+ * delay between scroll events, and we do not want to interrupt a scroll with a
+ * render, it cannot be too small. 200ms seems to be around the lower limit for
+ * this value on Android / iOS.
+ */
+const RESET_SCROLL_REFERENCE_POINT_WAIT_MS = 200;
+
+/**
+ * @param {!JsonObject} props
+ * @return {PreactDef.Renderable}
+ */
+export function Scroller(props) {
+  const {
+    children,
+    ignoreProgrammaticScroll,
+    loop,
+    restingIndex,
+    setRestingIndex,
+  } = props;
+
+  /**
+   * The number of slides we want to place before the
+   * reference or resting index. Only needed if loop=true.
+   */
+  const pivotIndex = Math.floor(children.length / 2);
+
+  /**
+   * The dynamic position that the slide at the resting index
+   * is with respect to its scrolling order. Only needed if loop=true.
+   */
+  const offsetRef = useRef(restingIndex);
+  const containerRef = useRef();
+  const slides = renderSlides({
+    children,
+    loop,
+    offsetRef,
+    pivotIndex,
+    restingIndex,
+  });
+  const currentIndex = useRef(restingIndex);
+
+  // Note: this has to be useLayoutEffect, not useEffect.
+  useLayoutEffect(() => {
+    if (!containerRef.current) {
+      return;
+    }
+    const container = containerRef.current;
+    ignoreProgrammaticScroll.current = true;
+    container./* OK */ scrollLeft = loop
+      ? container./* OK */ offsetWidth * pivotIndex
+      : container./* OK */ offsetWidth * restingIndex;
+  }, [ignoreProgrammaticScroll, loop, restingIndex, pivotIndex]);
+
+  // Trigger render by setting the resting index to the current scroll state.
+  const resetScrollReferencePoint = () => {
+    // Check if the resting index we are centered around is the same as where
+    // we stopped scrolling. If so, we do not need to move anything.
+    if (currentIndex.current === restingIndex) {
+      return;
+    }
+    setRestingIndex(currentIndex.current);
+  };
+  const debouncedResetScrollReferencePoint = debounce(
+    window,
+    resetScrollReferencePoint,
```

Comment:
It looks like we should instead inline `resetScrollReferencePoint` here and make the whole `debouncedResetScrollReferencePoint` to use `useMemo` so that we only have one debouncer for the life-time of the scroller.

Reply:
Let's dedup with [this thread](https://github.com/ampproject/amphtml/pull/28704#discussion_r436193014) -- I think the case still stands that because `resetScrollReferencePoint` uses `restingIndex` and `setRestingIndex` declared from `useState`, that would have to be in the dependency array for `useMemo` or `useCallback` -- but because the call to `setRestingIndex` changes the value of `restingIndex`, won't this function have to be redefined on every render anyway? 

I suppose we could make a `restingIndexRef` whose `current` is always updated to the value of `restingIndex` and compare that in `resetScrollReferencePoint` instead.

Reply:
BTW I did make the change you recommended so we can discuss it more concretely.